File: Enhancement One New/TaskService.py
import random
import Task


class TaskService:
    def __init__(self):
        self.task_list = []
        self.task_count = 0

    # list of tasks
    def get_task_list(self):
        return self.task_list

    # count tasks
    def get_task_count(self):
        return self.task_count

    # print
    def print_task_list(self):
        for task in self.task_list:
            print("Task ID: " + task.get_id() + " Name: " + task.get_task_name())
            print("Description: " + task.get_task_description() + "\n")

    # add task
    def add_task(self, p_name, p_description):
        if p_name is None or len(p_name) > 20:
            raise ValueError("Invalid Name")
        if p_description is None or len(p_description) > 50:
            raise ValueError("Invalid Description")

        new_id = self.generate_unique_id()
        new_task = Task(new_id, p_name, p_description)
        self.task_list.append(new_task)
        self.task_count += 1
        print("Task " + new_task.get_task_name() + " created and added successfully")

    def remove_task_instance(self, task):
        self.task_list.remove(task)
        self.task_count -= 1

    def remove_task(self, task_id):
        if not self.task_list:
            print("There are no tasks to remove")
            return

        if task_id is None or len(task_id) > 10:
            raise ValueError("Invalid ID")

        index = -1
        for idx, task in enumerate(self.task_list):
            if task.get_id() == task_id:
                index = idx
                break

        if index == -1:
            print("Task not found")
            return
        else:
            self.remove_task_instance(self.task_list[index])
            print("Task was removed from the list")

    def update_task(self, task_id, update, selection):
        if task_id is None or len(task_id) > 10 or update is None or selection < 0:
            raise ValueError("Invalid Request")

        if not self.task_list:
            print("There are no tasks to update")
            return

        index = -1
        for idx, task in enumerate(self.task_list):
            if task.get_id() == task_id:
                index = idx
                break

        if index == -1:
            print("Task not found")
            return

        update_task = self.task_list[index]

        # conditional for updates
        # 1 is for name
        # 2 is for description
        if selection == 1:
            update_task.set_task_name(update)
        elif selection == 2:
            update_task.set_task_desc(update)
        else:
            print("Task not updated, error")
            return

        self.remove_task_instance(self.task_list[index])
        self.task_list.append(update_task)

    def generate_unique_id(self):
        rand = random.Random()
        new_id = rand.randint(0, 999999999)
        unique_id = str(new_id)

        id_list = [task.get_id() for task in self.task_list]
        while unique_id in id_list or len(unique_id) > 10:
            new_id = rand.randint(0, 999999999)
            unique_id = str(new_id)

        print("New Task ID created:", unique_id)
        return unique_id

File: Enhancement One Original/projectpackage/TaskService.java
package projectpackage;

import java.util.Random;
import java.util.Vector;

public class TaskService {

	private Vector<Task> taskList = new Vector<Task>();
	private int taskCount = 0;
	
	//list of tasks
	public Vector<Task> GetTaskList() {
		return taskList;
	}
	//count tasks
	public int GetTaskCount() {
		return taskCount;
	}
	//print
	public void PrintTaskList() {
		for (Task t : taskList) {
			System.out.println("Task ID: " + t.GetID() + " Name: " + t.GetTaskName());
			System.out.println("Description: " + t.GetTaskDescription() + "\n");
		}
	}
	//add task
	public void AddTask(String pName, String pDescription) {
		if (pName == null || pName.length() > 20) {
			throw new IllegalArgumentException("Invalid Name");
		}
		if (pDescription == null || pDescription.length() > 50) {
			throw new IllegalArgumentException("Invalid Description");
		}
		String newID = GenerateUniqueID();
		var newTask = new Task(newID, pName, pDescription);
		taskList.add(newTask);
		taskCount = GetTaskCount() + 1;
		System.out.println("Task " + newTask.GetTaskName() + " created and added successfully");
	}
	private void AddTask(Task task) {
		taskList.add(task);
		taskCount = GetTaskCount() + 1;
	}
	private void RemoveTask(Task task) {
		taskList.remove(task);
		taskCount = GetTaskCount() - 1;
	}
	
	public void RemoveTask(String id) {
		if (taskList.isEmpty()) {
			System.out.println("There are no tasks to remove");
		}
		if (id == null || id.length() > 10) {
			throw new IllegalArgumentException("Invalid ID");
		}
		int index = -1;
		for (Task t : taskList) {
			if (t.GetID() == id) {
				index = taskList.indexOf(t);
			}
		}
		if (index == -1) {
			System.out.println("Task not found");
			return;
		}
		else {
			taskList.remove(index);
			taskCount = GetTaskCount() - 1;
			System.out.println("Task was removed from the list");
		}
	}
	
	public void UpdateTask(String id, String update, int selection) {
		if (id == null || id.length() > 10 || update == null || selection < 0) {
			throw new IllegalArgumentException("Invalid Request");
		}
		if (taskList.isEmpty()) {
			System.out.println("There are no tasks to update");
			return;
		}
		int index = -1;
		for (Task t : taskList) {
			if (t.GetID() == id) {
				index = taskList.indexOf(t);
			}
		}
		if (index == 1) {
			System.out.println("Task not found");
			return;
		}
		Task updateTask = taskList.get(index);
		//conditional for updates
		// 1 is for name
		// 2 is for description
		switch(selection) {
		case 1: {
			updateTask.SetTaskName(update);;
			break;
		}
		case 2: {
			updateTask.SetTaskDesc(update);;
			break;
		}
		default: {
			System.out.println("Task not updated, error");
			break;
		}
		}
		RemoveTask(taskList.elementAt(index));
		AddTask(updateTask);
	}
	
	public String GenerateUniqueID() {
		Random rand = new Random();
		int newID = rand.nextInt(1000000000);
		String uniqueID = Integer.toString(newID);
		
		Vector<String> idList = new Vector<String>();
		for (Task t : taskList) {
			idList.add(t.GetID());
		}
		while (idList.contains(uniqueID) || uniqueID.length() > 10) {
			newID = rand.nextInt(1000000000);
			uniqueID = Integer.toString(newID);
		}
		idList = null;
		return uniqueID;
	}
	
}

